File: metactical/custom_scripts/material_request/material_request.py
import frappe
from erpnext.stock.doctype.material_request.material_request import MaterialRequest
from erpnext.stock.utils import get_stock_balance
import json

class CustomMaterialRequest(MaterialRequest):
	def before_submit(self):
		super(CustomMaterialRequest, self).before_submit()
		self.validate_item_qty()
		self.validate_qoh()

	def validate_item_qty(self):
		if self.material_request_type == 'Material Transfer':
			for item in self.items:
				if item.from_warehouse is None or item.from_warehouse == '':
					frappe.throw(f'Source Warehouse is required for item {item.item_code}')

				# if item.qty > get_stock_balance(item.item_code, item.from_warehouse):
				# 	frappe.throw(f'Requested quantity of item {item.item_code} is greater than available quantity at warehouse')

	def validate_qoh(self):
		# check if a qty greater than the quantity on hand is entered
		if self.material_request_type == "Material Transfer":
			items = []
			for item in self.items:
				if item.qty > item.qoh:
					items.append(item.item_code) 

			if items:
				message = 'The quantities of these items exceed the available warehouse stock. <br>'
				message += "<b>" + ', '.join(items) + "</b>"
				frappe.msgprint(message)

	def before_save(self):
		self.set_status(update=True)	
		if len(self.items) > 0:
			self.set_default_supplier()
			suppliers = ''
			unique_suppliers = []
			for item in self.items:
				if item.ais_default_supplier and item.ais_default_supplier not in unique_suppliers:
					unique_suppliers.append(item.ais_default_supplier)
					if suppliers != '':
						suppliers += ', ' + item.ais_default_supplier
					else:
						suppliers += item.ais_default_supplier
			self.ais_suppliers = suppliers

	def set_default_supplier(self):
		for item in self.items:
			if not item.ais_default_supplier:
				doc = frappe.get_doc('Item', item.item_code)
				for defaults in doc.item_defaults:
					if defaults.company == self.company:
						item.ais_default_supplier = defaults.default_supplier

def set_default_supplier(self):
	for item in self.items:
		if not item.ais_default_supplier:
			doc = frappe.get_doc('Item', item.item_code)
			for defaults in doc.item_defaults:
				if defaults.company == self.company:
					item.ais_default_supplier = defaults.default_supplier

@frappe.whitelist()
def get_target_warehouse(doctype, txt, searchfield, start, page_len, filters):
	user = filters.get("user")
	warehouses = []
	if user:
		setting_exists = frappe.db.get_value("Material Request User Permission", filters={"user": user})
		if setting_exists:
			warehouses = frappe.db.sql("""SELECT warehouse FROM `tabMaterial Request Permitted Warehouse` 
							WHERE warehouse LIKE %(txt)s AND parent= %(parent)s
							AND parentfield='permitted_target_warehouse'""", 
							{
								'txt': "%%%s%%" % txt,
								'parent': setting_exists
							})
		else:
			#Retrun all warehouses
			warehouses = frappe.db.sql("""SELECT name FROM `tabWarehouse` WHERE is_group=0 AND disabled=0 AND name LIKE %(txt)s""", {'txt': "%%%s%%" % txt})
	return warehouses

@frappe.whitelist()
def get_qoh(filters):
	filters = json.loads(filters)
	updated_qty = []

	for value in filters:
		qty = 0
		data= frappe.db.sql("""select actual_qty-reserved_qty AS qty from `tabBin`
			where item_code = %s and warehouse=%s
			""",(value['item'], value["warehouse"]), as_dict=1)
		if data and data[0]['qty'] > 0:
			qty = data[0]['qty']
		
		updated_qty.append({
			'item_code': value['item'],
			'qty': qty,
			'name': value['name']
		})

	return updated_qty

File: metactical/custom_scripts/material_request/material_request.js
frappe.ui.form.on('Material Request', {
	refresh: function(frm) {
		if (frm.doc.docstatus == 0) {
			get_quantities_on_hand(frm);
		}
	},
	onload: function(frm) {
		frappe.after_ajax(function(){
			frm.set_query("set_warehouse", function(doc){
				return {
					query: "metactical.custom_scripts.material_request.material_request.get_target_warehouse",
					filters: {"user": frappe.session.user}
				}
			});

			frm.set_query( "warehouse", "items", function(){
				return {
					query: "metactical.custom_scripts.material_request.material_request.get_target_warehouse",
					filters: {"user": frappe.session.user}
				}
			});
		});
	},

	get_item_data: function(frm, item, overwrite_warehouse=false) {
		// Metactical customizatino: Prevent overwriting of target warehouse
		overwrite_warehouse = false;
		
		if (item && !item.item_code) { return; }
		frm.call({
			method: "erpnext.stock.get_item_details.get_item_details",
			child: item,
			args: {
				args: {
					item_code: item.item_code,
					warehouse: item.warehouse,
					from_warehouse: item.from_warehouse,
					doctype: frm.doc.doctype,
					buying_price_list: frappe.defaults.get_default('buying_price_list'),
					currency: frappe.defaults.get_default('Currency'),
					name: frm.doc.name,
					qty: item.qty || 1,
					stock_qty: item.stock_qty,
					company: frm.doc.company,
					conversion_rate: 1,
					material_request_type: frm.doc.material_request_type,
					plc_conversion_rate: 1,
					rate: item.rate,
					uom: item.uom,
					conversion_factor: item.conversion_factor
				},
				overwrite_warehouse: overwrite_warehouse
			},
			callback: function(r) {
				const d = item;
				const qty_fields = ['actual_qty', 'projected_qty', 'min_order_qty'];
				
				if(!r.exc) {
					$.each(r.message, function(k, v) {
						if(!d[k] || in_list(qty_fields, k)) d[k] = v;
					});
				}
				
				//For default supplier
				d.ais_default_supplier = r.message['supplier'];
			}
		});
	},
	
});

frappe.ui.form.on('Material Request Item', {
	qty: function(frm, cdt, cdn) {
		var row = locals[cdt][cdn];
		if (row.qty > row.qoh) {
			frappe.msgprint(__("Quantity cannot be greater than Quantity on Hand"));
			row.qty = row.qoh;
			frm.refresh_field('items');
		}
	},

	item_code: function(frm, cdt, cdn) {
		get_qoh(frm, cdt, cdn)
	},

	from_warehouse: function(frm, cdt, cdn) {
		// frm.events.get_item_data(frm, item, true);
		get_qoh(frm, cdt, cdn);
	},
});

var get_quantities_on_hand = function(frm) {
	if (frm.doc.items) {
		var rows = []
		$.each(frm.doc.items, function(i, d) {
			if (d.item_code && d.from_warehouse) {
				rows.push({"item": d.item_code, "warehouse": d.from_warehouse, "name": d.name});
			}
		});

		if (rows.length > 0) {
			get_qoh(frm, null, null, rows);
		}
	}
}

var get_qoh = function(frm, cdt=null, cdn=null, rows=null) {
	if (!rows){
		var row = locals[cdt][cdn];
		if (!row.item_code || !row.from_warehouse) {
			return
		}
		else{
			rows = [{
				"name": row.name,
				"item": row.item_code,
				"warehouse": row.from_warehouse
			}]
		}
	}


	frappe.call({
		method: "metactical.custom_scripts.material_request.material_request.get_qoh",
		args: {
			"filters": rows
		},
		callback: function(r) {
			var rows = r.message;

			$.each(rows, function(i, d) {
				var row = locals['Material Request Item'][d.name];
				row.qoh = d.qty;
			});

			frm.refresh_field('items');
		}
	});
}